lib/internal/modules/esm/importmap.js

```diff
@@ -0,0 +1,135 @@
+'use strict';
+const path = require('path');
+const { readFileSync } = require('fs');
+const { pathToFileURL, isURL } = require('internal/url');
+const { setOwnProperty } = require('internal/util');
+const { JSONParse, ObjectEntries, SafeMap } = primordials;
+// @TODO add error codes
+// const { } = require('internal/errors').codes;
+
+class ImportMap {
+  #baseURL;
+  imports = new SafeMap();
+  scopes = new SafeMap();
+
+  constructor(raw, baseURL) {
+    this.#baseURL = baseURL;
+    processImportMap(this, baseURL, raw);
+  }
+
+  get baseURL() {
+    return this.#baseURL;
+  }
+
+  static load(filepath) {
+    let str;
+    try {
+      str = readFileSync(filepath);
+    } catch (err) {
+      if (err.code !== 'ENOENT') {
+        setOwnProperty(err, 'importMapPath', `${filepath}`);
+        throw err;
+      }
+      return;
+    }
+
+    let json;
+    try {
+      json = JSONParse(str);
+    } catch (err) {
+      setOwnProperty(err, 'importMap', `${str}`);
+      throw err;
+    }
+
+    return new ImportMap(json, pathToFileURL(path.dirname(filepath)));
+  }
+
+  resolve(specifier, parentURL) {
+    const scope = parentURL || this.baseURL;
+
+    // Process scopes
+    // @TODO this almost certainly needs to be a radix tree
+    for (const [prefix, mapping] of this.scopes) {
+      let mappedSpecifier = mapping.get(specifier);
+      if (scope.pathname.startsWith(prefix.pathname) && mappedSpecifier) {
+        if (!isURL(mappedSpecifier)) {
+          mappedSpecifier = pathToFileURL(mappedSpecifier, this.baseURL.pathname);
+          mapping.set(specifier, mappedSpecifier);
+        }
+        specifier = mappedSpecifier;
+        break;
+      }
+    }
+
+    let spec = specifier;
+    if (isURL(specifier)) {
+      spec = specifier.pathname;
+    }
+    let importMapping = this.imports.get(spec);
+    if (importMapping) {
+      if (!isURL(importMapping)) {
+        importMapping = pathToFileURL(importMapping, this.baseURL.pathname);
+        this.imports.set(spec, importMapping);
+      }
+      return importMapping;
+    }
+
+    return specifier;
+  }
+}
+
+function processImportMap(importMap, baseURL, raw) {
+  // validation and normalization
+  // validate basic structure
+  // sort the scopes by specificity
+  if (typeof raw.imports !== 'object') {
```

Comment:
My worry is not about someone pulling a fast one, it's about unexpected data breaking the implementation: Arrays exist in json. I believe this would pass what you have:

```json
{
  "imports": [
    ["foo", "../foo.mjs"]
  ]
}
```

And then blow up unexpectedly/ungracefully.

I guess here we only need to check for Array, so `typeof raw.imports !== 'object' || !Array.isArray(raw.imports)` probably covers it?